src/workerd/jsg/commonjs.c++

```diff
@@ -0,0 +1,163 @@
+      exports(js.v8Ref(module->getExports(js))) {}
+
+v8::Local<v8::Value> NodeJsModuleContext::require(jsg::Lock& js, kj::String specifier) {
+  // If it is a bare specifier known to be a Node.js built-in, then prefix the
+  // specifier with node:
+  bool isNodeBuiltin = false;
+  auto resolveOption = jsg::ModuleRegistry::ResolveOption::DEFAULT;
+  KJ_IF_SOME(spec, checkNodeSpecifier(specifier)) {
+    specifier = kj::mv(spec);
+    isNodeBuiltin = true;
+    resolveOption = jsg::ModuleRegistry::ResolveOption::BUILTIN_ONLY;
+  }
+
+  // TODO(cleanup): This implementation from here on is identical to the
+  // CommonJsModuleContext::require. We should consolidate these as the
+  // next step.
+
+  auto modulesForResolveCallback = jsg::getModulesForResolveCallback(js.v8Isolate);
+  KJ_REQUIRE(modulesForResolveCallback != nullptr, "didn't expect resolveCallback() now");
+
+  kj::Path targetPath = ([&] {
+    // If the specifier begins with one of our known prefixes, let's not resolve
+    // it against the referrer.
+    if (specifier.startsWith("node:") || specifier.startsWith("cloudflare:") ||
+        specifier.startsWith("workerd:")) {
+      return kj::Path::parse(specifier);
+    }
+    return path.parent().eval(specifier);
+  })();
+
+  // require() is only exposed to worker bundle modules so the resolve here is only
+  // permitted to require worker bundle or built-in modules. Internal modules are
+  // excluded.
+  auto& info =
+      JSG_REQUIRE_NONNULL(modulesForResolveCallback->resolve(js, targetPath, path, resolveOption,
+                              ModuleRegistry::ResolveMethod::REQUIRE, specifier.asPtr()),
+          Error, "No such module \"", targetPath.toString(), "\".");
+  // Adding imported from suffix here not necessary like it is for resolveCallback, since we have a
+  // js stack that will include the parent module's name and location of the failed require().
+
+  if (!isNodeBuiltin) {
+    JSG_REQUIRE_NONNULL(
+        info.maybeSynthetic, TypeError, "Cannot use require() to import an ES Module.");
+  }
+
+  return ModuleRegistry::requireImpl(js, info, ModuleRegistry::RequireImplOptions::EXPORT_DEFAULT);
+}
+
+v8::Local<v8::Value> NodeJsModuleContext::getBuffer(jsg::Lock& js) {
+  auto value = require(js, kj::str("node:buffer"));
+  JSG_REQUIRE(value->IsObject(), TypeError, "Invalid node:buffer implementation");
+  auto module = value.As<v8::Object>();
+  auto buffer = js.v8Get(module, "Buffer"_kj);
+  JSG_REQUIRE(buffer->IsFunction(), TypeError, "Invalid node:buffer implementation");
+  return buffer;
+}
+
+v8::Local<v8::Value> NodeJsModuleContext::getProcess(jsg::Lock& js) {
+  auto value = require(js, kj::str("node:process"));
+  JSG_REQUIRE(value->IsObject(), TypeError, "Invalid node:process implementation");
+  return value;
+}
+
+kj::String NodeJsModuleContext::getFilename() {
+  return path.toString(true);
+}
+
+kj::String NodeJsModuleContext::getDirname() {
+  return path.parent().toString(true);
```

Comment:
What if parent does not exist, and we are on root? How does this behave?

Reply:
I'm actually hoping to be able to remove this entire class soon so for now I don't want to make actual changes to this implementation. More changes are coming soon. I'd prefer to keep this focused only on moving the existing impl